Userland/Games/BrickGame/BrickGame.cpp

```diff
@@ -0,0 +1,554 @@
+    }
+
+    Block& random_shape()
+    {
+        m_shape = get_random_uniform(number_of_shapes);
+        m_rotation = 0;
+        m_position = { 6, 0 };
+        return *this;
+    }
+
+    [[nodiscard]] bool operator[](Position pos) const
+    {
+        return (block_row(pos.y() - m_position.y()) & (1 << (31 - pos.x()))) != 0;
+    }
+
+    [[nodiscard]] bool has_collision(Well const& well) const
+    {
+        for (size_t start_row = 0; start_row < shape_size; ++start_row) {
+            auto const row_index = start_row + m_position.y();
+            if (row_index >= Well::rows() || (well[row_index] & block_row(start_row)) != 0)
+                return true;
+        }
+        return false;
+    }
+
+    void place_into(Well& well)
+    {
+        for (size_t row_index = 0; row_index < shape_size; ++row_index)
+            well[m_position.y() + row_index] |= block_row(row_index);
+    }
+
+    [[nodiscard]] bool dot_at(Position position) const
+    {
+        return ((shape_data_at(position.y()) & (1 << (3 - position.x()))) != 0);
+    }
+
+private:
+    static constexpr u8 number_of_shapes = 7;
+    static constexpr u8 number_of_rotations = 4;
+
+    using Shape = u16;
+    using Row = u32;
+
+    static constexpr Shape s_shapes[number_of_shapes][number_of_rotations] = {
+        // Shape: I
+        { 0b0000'1111'0000'0000, 0b0010'0010'0010'0010, 0b0000'1111'0000'0000,
+            0b0010'0010'0010'0010 },
+
+        // Shape: J
+        { 0b0000'0111'0001'0000, 0b0001'0001'0011'0000, 0b0000'0100'0111'0000,
+            0b0011'0010'0010'0000 },
+
+        // Shape: L
+        { 0b0000'0111'0100'0000, 0b0110'0010'0010'0000, 0b0000'0001'0111'0000,
+            0b0010'0010'0011'0000 },
+
+        // Shape: O
+        { 0b0000'0110'0110'0000, 0b0000'0110'0110'0000, 0b0000'0110'0110'0000,
+            0b0000'0110'0110'0000 },
+
+        // Shape: S
+        { 0b0000'0011'0110'0000, 0b0100'0110'0010'0000, 0b0000'0011'0110'0000,
+            0b0100'0110'0010'0000 },
+
+        // Shape: T
+        { 0b0000'0111'0010'0000, 0b0001'0011'0001'0000, 0b0000'0010'0111'0000,
+            0b0100'0110'0100'0000 },
+
+        // Shape: Z
+        { 0b0000'0110'0011'0000, 0b0001'0011'0010'0000, 0b0000'0110'0011'0000,
+            0b0001'0011'0010'0000 }
+    };
+
+    Position m_position {};
+    u8 m_rotation {};
+    u8 m_shape {};
+
+    [[nodiscard]] Row block_row(size_t row) const
+    {
+        return shape_data_at(row) << (32 - m_position.x() - shape_size);
+    }
+
+    [[nodiscard]] Row shape_data_at(size_t row) const
+    {
+        switch (row) {
+        case 0:
+            return Row((s_shapes[m_shape][m_rotation] >> 12) & 0xf);
+        case 1:
+            return Row((s_shapes[m_shape][m_rotation] >> 8) & 0xf);
+        case 2:
+            return Row((s_shapes[m_shape][m_rotation] >> 4) & 0xf);
+        case 3:
+            return Row(s_shapes[m_shape][m_rotation] & 0xf);
+        default:
+            return Row {};
```

Comment:
Is this an impossible scenario within normal execution? Normally we would then just `VERIFY_NOT_REACHED`, but if a default-constructed object is intended and used then it's fine

Reply:
Here it actually need to returns an empty row, i.e. filles with zeroes. It simplifies shape boundary checking.